ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_mvmv ?log obj v u)
+    in
+    let r =
+      match a, b with
+      | Amode left, Amode right -> submode_cc left right
+      | Amodevar v, Amode right -> submode_mvc v right
+      | Amode left, Amodevar v -> submode_cmv left v
+      | Amodevar v, Amodevar u -> submode_mvmv v u
+      | Amode a, Amodemeet (b, mvs) ->
+        Result.bind (submode_cc a b) (fun () ->
+            find_error (fun mv -> submode_cmv a mv) mvs)
+      | Amodevar mv, Amodemeet (b, mvs) ->
+        Result.bind (submode_mvc mv b) (fun () ->
+            find_error (fun mv' -> submode_mvmv mv mv') mvs)
+      | Amodejoin (a, mvs), Amode b ->
+        Result.bind (submode_cc a b) (fun () ->
+            find_error (fun mv' -> submode_mvc mv' b) mvs)
+      | Amodejoin (a, mvs), Amodevar mv ->
+        Result.bind (submode_cmv a mv) (fun () ->
+            find_error (fun mv' -> submode_mvmv mv' mv) mvs)
+      | Amodejoin (a, mvs), Amodemeet (b, mus) ->
+        (* TODO: mabye create a intermediate variable? *)
+        Result.bind (submode_cc a b) (fun () ->
+            Result.bind
+              (find_error (fun mv -> submode_mvc mv b) mvs)
+              (fun () ->
+                Result.bind
+                  (find_error (fun mu -> submode_cmv a mu) mus)
+                  (fun () ->
+                    find_error
+                      (fun mu -> find_error (fun mv -> submode_mvmv mv mu) mvs)
+                      mus)))
+    in
+    match r with
+    | Ok () -> Ok log
+    | Error e ->
+      (* we mutated some states and found conflict;
+         need to revert those mutation to keep the state consistent.
+         A nice by-effect is that this function doesn't mutate state in failure
+      *)
+      Option.iter (fun log -> undo_changes !log) log;
+      Error e
+
+  let submode obj a b =
+    match submode_try obj a b with
+    | Ok log ->
+      Option.iter !append_changes log;
+      Ok ()
+    | Error _ as e -> e
+
+  let zap_to_ceil_morphvar obj mv =
+    assert (submode obj (Amode (mupper obj mv)) (Amodevar mv) |> Result.is_ok);
+    mupper obj mv
+
+  let zap_to_ceil : type a l. a C.obj -> (a, l * allowed) mode -> a =
+   fun obj -> function
+    | Amode m -> m
+    | Amodevar mv -> zap_to_ceil_morphvar obj mv
+    | Amodemeet (a, mvs) ->
+      List.fold_left
+        (fun acc mv -> C.meet obj acc (zap_to_ceil_morphvar obj mv))
+        a mvs
+
+  let join (type a r) obj l =
+    let rec loop :
+        a ->
+        (a, allowed * disallowed) morphvar list ->
+        (a, allowed * r) mode list ->
+        (a, allowed * disallowed) mode =
+     fun a mvs ->
+      if C.le obj (C.max obj) a
+      then fun _ -> Amode (C.max obj)
+      else
+        function
+        | [] -> Amodejoin (a, mvs)
+        | mv :: xs -> (
+          match disallow_right mv with
+          | Amode b -> loop (C.join obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
```

Comment:
De-duplicate?

Reply:
As suggested by the other comments, I think we will replace the `list` with patricia-tree in a future PR.